### PR TITLE
Akopyan Zal lab2 var2

### PR DESCRIPTION
**Pull request by akopyanzal**

### CODE REVIEW

File: llvm/labs/lab2/akopyan_zal/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+if (NOT WIN32 AND NOT CYGWIN)
+
+    file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.h *.hpp)
+    add_llvm_pass_plugin( ${Plugin_Name} ${ALL_SOURCE_FILES}
+    DEPENDS intrinsics_gen BUILDTREE_ONLY)
+    set(LLVM_TEST_DEPENDS ${Plugin_Name} ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
+
+    message(STATUS "Plugin: ${Plugin_Name} -- BUILDED")
+else()
+    message(STATUS "Plugin: ${Plugin_Name} -- NOT BUILDED")
```
Comment (aobolensk) [inline llvm/labs/lab2/akopyan_zal/CMakeLists.txt]:
```suggestion
    message(STATUS "Plugin: ${Plugin_Name} -- NOT BUILT")
```

File: llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        }
+      }
+    }
+  }
+
+  void removeCallInstruction(CallInst *CI) { CI->eraseFromParent(); }
+
+  void moveSplitBBAfterNextBlock(Function &F, BasicBlock *SplitBB) {
+    Function::iterator it = F.begin();
+    while (&*it != SplitBB && it != F.end()) {
```
Comment (aobolensk) [inline llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]:
```suggestion
    while (it != F.end() && &*it != SplitBB) {
```
Dangerous operation. You should check first and only if it is available you can dereference it.

File: llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        if (RI->getParent() != SplitBB) {
+          RI->eraseFromParent();
+          Builder.SetInsertPoint(&BB);
+          Builder.CreateBr(SplitBB);
+        }
+      }
+    }
+  }
+
+  void removeCallInstruction(CallInst *CI) { CI->eraseFromParent(); }
```
Comment (aobolensk) [inline llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]: Do we really need such wrapper?

File: llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    processCallSites(F, callsToInline);
+
+    return PreservedAnalyses::none();
+  }
+
+  static bool isRequired() { return true; }
+
+private:
+  std::vector<CallInst *> findCallsToInline(Function &F) {
+    Function *Callee{};
```
Comment (aobolensk) [inline llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]: Do we need to define `Callee` variable in the beginning of the function?

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp (hunk 4 above)
Comment (akopyanzal) [reply llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]: Yes, I think that would be appropriate here. If you don't create a Callee variable, you'll have to call CI->getCalledFunction() twice, which may be less efficient.

Comment (aobolensk) [reply llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]: I'm not sure why. On the line 38
```cpp
Callee = CI->getCalledFunction();
```
You do that unconditionally, so you override `Callee` variable on each iteration. What do you mean by calling `CI->getCalledFunction();` twice?

Comment (akopyanzal) [reply llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]: I don't quite understand what your question is. As far as I understand, you can write it in different ways (a kind of syntactic sugar). You can do this function without an explicit `Callee` declaration at the beginning of the function. If I misunderstand something, please explain the problem in more detail.

```cpp
  std::vector<CallInst *> findCallsToInline(Function &F) {
    std::vector<CallInst *> callsToInline{};
    for (auto &BB : F) {
      for (auto &I : BB) {
        if (auto *CI = dyn_cast<CallInst>(&I)) {
          if (Function *Callee = CI->getCalledFunction()) {
            if (Callee->arg_size() == 0 &&
                Callee->getReturnType()->isVoidTy()) {
              callsToInline.push_back(CI);
            }
          }
        }
      }
    }
    return callsToInline;
  }
```

Comment (aobolensk) [reply llvm/labs/lab2/akopyan_zal/akopyanInliningPass.cpp]: Yes, you can use it like in the code snippet you shared in the previous message, or just move `Function *Callee{};` inside `if (auto *CI = dyn_cast<CallInst>(&I)) {`.

My point is that there is no need to initialize `Callee` var in outer scope

**Review by aobolensk — APPROVED** [review]

**Pull request merged**